prow/gerrit/adapter/trigger.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"k8s.io/test-infra/prow/pjutil"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gerrit/client"
+)
+
+const layout = "2006-01-02 15:04:05"
+
+// messageFilter builds a filter for jobs based on the messageBody matching the trigger regex of the jobs.
+func messageFilter(lastUpdate time.Time, change client.ChangeInfo, presubmits []config.Presubmit) (pjutil.Filter, error) {
+	var filters []pjutil.Filter
+	currentRevision := change.Revisions[change.CurrentRevision].Number
+	for _, message := range change.Messages {
+		// Skip comments not germane to this plugin
+		if !pjutil.TestAllRe.MatchString(message.Message) {
+			matched := false
+			for _, presubmit := range presubmits {
+				matched = matched || presubmit.TriggerMatches(message.Message)
+				if matched {
+					break
+				}
+			}
+			if !matched {
+				logrus.Infof("Comment %s doesn't match any triggering regex, skipping.", message.Message)
+				continue
+			}
+
+		} else {
+			filters = append(filters, pjutil.TestAllFilter())
+		}
+		messageTime, err := time.Parse(layout, message.Date)
+		if err != nil {
+			logrus.WithError(err).Errorf("Parse time %v failed", message.Date)
+			continue
+		}
+		if message.RevisionNumber == currentRevision && messageTime.After(lastUpdate) {
+			filters = append(filters, pjutil.CommandFilter(message.Message))
+		}
+	}
+	return pjutil.AggregateFilter(filters), nil
+}
+
+// oldRevisionFilter builds a filter for jobs that have already been processed for a revision.
+func oldRevisionFilter(lastUpdate time.Time, rev client.RevisionInfo) pjutil.Filter {
+	created, err := time.Parse(layout, rev.Created)
+	if err != nil {
+		logrus.WithError(err).Errorf("Parse time %v failed", rev.Created)
+		return func(p config.Presubmit) (bool, bool, bool) {
+			return false, false, false
+		}
+	}
+
+	return func(p config.Presubmit) (bool, bool, bool) {
```

Comment:
Isn't this handled by the abort logic in controllers? Why should we care during triggering?

Reply:
Not when there's a new message.

Reply:
How is this different from GitHub when we have many comments? Just that you want to optimize and do it before triggering? Not sure if we super want this, would be confusing to someone debugging if the logic is in two places

Reply:
Yeah, this was bugging me as well. What we really want is to be able to get an updated revision from the client controller (but the gerrit api doesn't have this /shrug) . So what we have currently is batched changes that have new commits or new messages or both. 
IIUC we can dedup these in github because we have demux logic for individual events https://github.com/kubernetes/test-infra/blob/master/prow/hook/server.go#L74?

Reply:
wilder idea: We can make gerrit-change a CRD from the adapter (which can contain all necessary info for triggering), and write another controller, or reuse the adapter, to reconcile the CRD state and existing prowjob state and hope that can make this logic less convoluted?

Reply:
Also, yes to optimizing before triggering. we could just send through everything from the client and only filter in trigger, but then that wouldn't be optimal.

Reply:
@krzyzacy Sure :sweat_smile:. Will leave it to you as an optimization later :smile:, or I might take a stab at it as a follow up.

Reply:
please ignore me :jack_o_lantern: